### PR TITLE
Feat/http post

### LINKED ISSUE #49: feat/http-post
Handle a HTTP POST request.

### PR DESCRIPTION
**Pull request by gbudau**
Handle a HTTP POST request.

To make the review a bit easier, I've created another PR for the FormFile class, #66.

closes #49

### CODE REVIEW

File: srcs/app/http/HttpRequestHandler.cpp [hunk 1]
```diff
@@ -295,6 +348,10 @@ void	HttpRequestHandler::DoDelete_(const HttpRequest &request) {
 	raw_response_ = response.CreateResponseString();
 }
 
+bool	HttpRequestHandler::IsExecutable_(const std::string &full_path) const {
+	return access(full_path.c_str(), X_OK) == 0;
```
[the remark this answers is not in the document]
Comment (gbudau) [reply srcs/app/http/HttpRequestHandler.cpp:352]: Good to know, thank you. Then should we remove this function and deal with this case when calling execve?

File: srcs/app/http/HttpRequestHandler.cpp [hunk 2]
```diff
@@ -271,16 +274,64 @@ void	HttpRequestHandler::DoGet_(const HttpRequest &request) {
...
+		} else {
+			// TODO(any) Implement CGI
+			RequestError_(501);
+		}
+	} else {
+		if (!IsUploadEnabled_() || !IsValidUploadPath_(request_path)) {
+			RequestError_(403);
+		} else {
+			UploadFile_(request);
+		}
```
Comment (vdedios) [inline srcs/app/http/HttpRequestHandler.cpp:333]: Think possitively, be optimistic! :smile: 

```suggestion
		if (IsUploadEnabled_() && IsValidUploadPath_(request_path)) {
			UploadFile_(request);
		} else {
			RequestError_(403);
		}
```

File: srcs/app/http/HttpRequestHandler.cpp [hunk 3]
```diff
@@ -271,16 +274,64 @@ void	HttpRequestHandler::DoGet_(const HttpRequest &request) {
...
+							request_location_->common.root + request_path;
+	if (IsRegularFile_(full_path)) {
+		if (!IsCGI_(full_path)) {
+			RequestError_(404);
+		} else {
+			// TODO(any) Implement CGI
+			RequestError_(501);
+		}
+	} else {
+		if (!IsUploadEnabled_() || !IsValidUploadPath_(request_path)) {
```
Comment (vdedios) [inline srcs/app/http/HttpRequestHandler.cpp:329]: I think a simple `POST` version without upload to a file is missing. I know it does not make sense but this method does not always require to save into a file. We could compare behaviour with nginx, just a simple form without a cgi//upload, I could not find it in your test html pages.

File: srcs/app/http/HttpRequestHandler.cpp [hunk 4]
```diff
@@ -271,16 +274,64 @@ void	HttpRequestHandler::DoGet_(const HttpRequest &request) {
...
+		} else {
+			DefaultStatusResponse_(200);
+		}
+	}
+	catch (const std::exception &e) {
+		RequestError_(400);
+	}
+}
+
+void	HttpRequestHandler::DoPost_(const HttpRequest &request) {
```
[the remark this answers is not in the document]
Comment (gbudau) [reply srcs/app/http/HttpRequestHandler.cpp:317]: Sure, your call.

File: srcs/app/http/HttpRequestHandler.cpp [hunk 5]
```diff
@@ -271,16 +274,64 @@ void	HttpRequestHandler::DoGet_(const HttpRequest &request) {
...
+	const std::string request_path = request.GetPath();
+	const std::string full_path =
+							request_location_->common.root + request_path;
+	if (IsRegularFile_(full_path)) {
+		if (!IsCGI_(full_path)) {
+			RequestError_(404);
```
Comment (gbudau) [inline srcs/app/http/HttpRequestHandler.cpp:323]: Here 501 (Not Implemented)

File: srcs/app/http/HttpRequestHandler.cpp [hunk 6]
```diff
@@ -271,16 +274,64 @@ void	HttpRequestHandler::DoGet_(const HttpRequest &request) {
...
+		} else {
+			// TODO(any) Implement CGI
+			RequestError_(501);
+		}
+	} else {
+		if (IsUploadEnabled_() && IsValidUploadPath_(request_path)) {
+			UploadFile_(request);
+		} else {
+			RequestError_(403);
+		}
```
Comment (gbudau) [inline srcs/app/http/HttpRequestHandler.cpp:333]: Here return 404(Not found)

**Pull request merged**